### ISSUE TITLE
[Question] Combine subscription and pure event handlers

### ISSUE THREAD

**Issue by hermann-p** [cljfx/cljfx#116]
A great way to create frontends!
I created an example app, testing many features and progressed from basic displaying to getting react+reduce/re-frame like behaviour.
Extracting state slices from context works really well (although my linter tells me `fx/sub` is deprecated), and creating pure event handlers with `fx.opt/map-event-handler` is quite easy after a few years of developing react/re-frame :D 

However, combining both seems to be a bit tricky.

The base parts are

```clojure
(def *state (fx/create-context {...} cache/lru-cache-factory))

(defmulti handle-event :event/type) ;; plus pure event implementations `state -> modified state` 

(def wrapped-handler
  (-> handle-event
      (fx/wrap-co-effects {:state (fx/make-deref-co-effect *state)})
      (fx/wrap-effects {:state (fx/make-reset-effect *state)
                        :dispatch fx/dispatch-effect})))

(def renderer
  (fx/create-renderer
   :middleware (comp
                fx/wrap-context-desc
                (fx/wrap-map-desc #(root-view %)))
   :opts {:fx.opt/map-event-handler wrapped-handler
          :fx.opt/type->lifecycle #(or (fx/keyword->lifecycle %)
                                       (fx/fn->lifecycle-with-context %))}))
```

I think I have to modify the handler-wrapper, and maybe there's even a pre-defined `fx/...` function to use? The current implementation worked for the state being an atom, but not for updating the context.

**Comment by hermann-p** [comment]
After some playing around, I managed to combine both by adapting the wrapper. To improve error messages I created named functions for the co-effect and effect needed.

```clojure
(defn extract-state []
  (:cljfx.context/m (deref *state)))

(defn swap-state [state foo]
  (swap! *state fx/reset-context state))

(def wrapped-handler
  (-> handle-event
      (fx/wrap-co-effects {:state extract-state})
      (fx/wrap-effects {:state swap-state
                        :dispatch fx/dispatch-effect})))
```

**Comment by vlaaad** [comment]
Hi! `:cljfx.context/m` is implementation details and subject to change, please don't rely on it! 
If you want to have effects that change the context, I think it's better for event handlers to return the new context, i.e.:
```clj
(defn extract-state [] ;; same as fx/make-deref-co-effect
  (deref *state))

(defn reset-state [state _] ;; same as fx/make-reset-effect
  (reset! *state state))

(def wrapped-handler
  (-> handle-event
      (fx/wrap-co-effects {:state extract-state})
      (fx/wrap-effects {:state reset-state
                        :dispatch fx/dispatch-effect})))
```
Then your event handlers will call to `fx/swap-context` and `fx/reset-context`, e.g.:
```clj
(defmethod ::on-text-changed [{:keys [state fx/event]}]
  {:state (fx/swap-state state assoc :text event)})
```
I would also use `:fx/context` for effect/co-effect names to keep it consistent with components functions that receive context on that key.

Does it help? :)

**Comment by hermann-p** [comment]
Hi Vlaaad!

Thanks for you answer, especially the part about the key name.

Your example gets rid of explicitly needing to name the key for sure.
On the other hand, when doing it this way it introduces the need to repeatedly type `(fx/swap-context... )` for every handler. In addition the handlers can no longer be trivially unit-tested without wrapping test-inputs as contexts and derefing the results again, which is the exact reason why I want to use what you call "event handling on steroids" :D 

Is there a way to keep extracting and setting the state data from/to the context inside the wrapper?

Also, what do you mean by using `:fx/context` for (co-)effect names?

**Comment by vlaaad** [comment]
Hmm, if I was using contexts in the app, I would also prefer testing contexts, otherwise I would not be testing code that runs in the real app... Why do you want to test raw maps instead of contexts? Do you want to use contexts in views and raw maps in event handling?

> Also, what do you mean by using `:fx/context` for (co-)effect names?
```clj
(fx/wrap-co-effects {:fx/context extract-state})
```

Btw, what I ended up using in [Reveal](https://github.com/vlaaad/reveal) is returning functions as effects and not having state co-effect at all (semantically speaking — I turned out I don't need neither co-effects, nor effects), i.e.:
```clj
(defn swap-state [f _]
  (swap! *state fx/swap-context f))

(def wrapped-handler
  (-> handle-event
      (fx/wrap-effects {:state swap-state
                        :dispatch fx/dispatch-effect})))

(defmethod ::on-text-changed [{:keys [fx/event]}]
  {:state #(assoc % :text event)})                        
```
This is more reliable when concurrent state modifications happen: if you `deref` your state when co-effect is executed, the state might change by the time the effect is executed.

**Comment by hermann-p** [comment]
TLDR: Yes, I think being able to use pure data in event handlers and only apply the contexts for the views is what I prefer to do.

**Explanation**

I disagree with testing contexts and their updates on events as a whole. Maybe it's a testing philosophy thing?

In my opinion, I expect the contexts to work, it should be the library's task to ensure that. The author's the expert on that case, after all. So I would just select the axiom that "the context will work as expected, when it gets set with the updated data". Of course, there will be some test in which I ensure my wrapper extracts and updates the state from the context correctly, but I'd do that only once (ok, maybe in some combinations...).

What I want to do is test my code on as small a margin as possible. When I want to test if an update with a complex calculation returns the correct result, I don't want to take additional steps to ensure that data is extracted, that data is transformed, and then the result gets stored. That's three things to test I'd like to isolate from each other.
Maybe the event event returns some new effect, like requirement to load something via http. Then I don't want to await that request every time it gets created (suddenly needing network connectivity or mocking for the test) but just want to see that the effect gets listed. I can test the effect instructions in a dedicated test.

**Done explaining**

I understand the concurrency problem. I had thought the event-generation would be instantly -- `(set-result (update (deref context)))`, but of course javafx is highly concurrent. Your suggested method is actually quite good for the use case -- it avoids coupling the event handlers to the state implementation, avoids concurrency issues, and allows to test the update logic by extracting the `:state` prop and throwing data against it.

I think that finally resolves my question.

**Comment by vlaaad** [comment]
Good to hear 👍

**Comment by hermann-p** [comment]
Thanks for your help :)
